lib/core/proxy.c

```diff
@@ -671,6 +677,22 @@ void h2o__proxy_process_request(h2o_req_t *req)
 
     struct rp_generator_t *self = proxy_send_prepare(req);
 
+    if ((sslen = req->conn->callbacks->get_peername(req->conn, (void *)&ss)) != 0) {
+        remote_addr_len = h2o_socket_getnumerichost((void *)&ss, sslen, remote_addr);
+        port = h2o_socket_getport((void *)&ss);
+    }
+    
+    /* if remote addr cannot be fetched, use a default one */
+    if (remote_addr_len == SIZE_MAX) {
+        strcpy(remote_addr, "169.254.0.1");
+        remote_addr_len = strlen(remote_addr);
+        port = 1000;
+    }
+
+    self->lb_req_info.path = req->path;
+    self->lb_req_info.port = port;
+    memcpy(self->lb_req_info.remote_addr, remote_addr, remote_addr_len);
+    self->lb_req_info.remote_addr_len = remote_addr_len;
```

Comment:
The information required for selecting the target depends on the balancer being selected. Or if balancer is not used, then there is no need for any information. In other words, proactively calculating the information here is useless in some configurations.

Considering that, could we let the balancer pull the information (by calling a callback) instead of pushing them? For example, would it be possible to create a callback in the configurator and pass it as an argument to `h2o_balancer_create_xxx`?

Reply:
I'll try to find a better way for pulling. Though it is sort of unclear about how to identify those information related to the original request by only information related to connection to an upstream.

Reply:
And it would be challenging if reproxy is on and balancer use something like hashing based on request path so that better locality could be expected. When the request is proceeded again because of reproxy, path should be changed. Then the required path information should be updated for next target determination.

Reply:
>  Though it is sort of unclear about how to identify those information related to the original request by only information related to connection to an upstream.

Pointer to `st_h2o_http1client_private_t` is passed as an argument to `h2o_socketpool_connect`. `st_h2o_http1client_private_t::super.data` is a pointer to `rp_generator_t`. The request that initiated the call to `h2o_socketpool_connect` is available as `rp_generator_t::src_req`.

So it should be possible to define a callback that resolves `src_req` from the arguments passed to `h2o_socketpool_connect` and returns a hash value. How you would let applications create such callback (without layer violation) would be an interesting question.

Reply:
> So it should be possible to define a callback that resolves src_req from the arguments passed to h2o_socketpool_connect and returns a hash value.

I know it could be possible but still afraid of future maintainance. OTOH it is not that clear to resolve `src_req` from arguments passed to `h2o_socketpool_connect` and I'm afraid that the resolving could become a barrier for future updates. I think it should be considered more carefully.

And since currently hash balancer is removed from this PR, we could simply remove this part, which is not required.

Reply:
> And since currently hash balancer is removed from this PR, we could simply remove this part, which is not required.

Thank you for pointing that out. Let's do that. It is always a good idea to remove code and structure that is not used at the moment.